gtsfm/utils/features.py

```diff
@@ -45,5 +45,66 @@ def normalize_coordinates(coordinates: np.ndarray, intrinsics: Cal3Bundler) -> n
     Returns:
         normalized coordinates, of shape Nx2.
     """
-
     return np.vstack([intrinsics.calibrate(x[:2].reshape(2, 1)) for x in coordinates])
+
+
+def convert_to_homogenous_coordinates(
+    non_homogenous_coordinates: np.ndarray,
+) -> np.ndarray:
+    """Convert coordinates to homogenous system (by appending a column of ones).
+
+    Args:
+        non_homogenous_coordinates: 2d non-homogenous coordinates, of shape Nx2.
+
+    Returns:
+        2d homogenous coordinates, of shape Nx3.
+
+    Raises:
+        TypeError: if input is not 2 dimensional.
+    """
+    if non_homogenous_coordinates is None or non_homogenous_coordinates.size == 0:
+        return non_homogenous_coordinates
+
+    if non_homogenous_coordinates.shape[1] != 2:
+        raise TypeError("Input should be 2D")
+
+    return np.hstack(
+        (
+            non_homogenous_coordinates,
+            np.ones((non_homogenous_coordinates.shape[0], 1)),
+        )
+    )
+
+
+def convert_to_epipolar_lines(normalized_coordinates_i1: np.ndarray, i2Ei1: EssentialMatrix) -> np.array:
+    """Convert coordinates to epipolar lines.
+
+    Args:
+        normalized_coordinates_i1: normalized coordinates in i1, of shape Nx2.
+        i2Ei1: essential matrix.
+
+    Returns:
+        Corr. epipolar lines in i2, of shape Nx3
+    """
+    if normalized_coordinates_i1 is None or normalized_coordinates_i1.size == 0:
+        return normalized_coordinates_i1
+
+    epipolar_lines = convert_to_homogenous_coordinates(normalized_coordinates_i1) @ i2Ei1.matrix().T
+
```

Comment:
nit: prefer to remove this extra whitespace here

Reply:
done